### PR TITLE
[RFC] Add support for Pipfiles

### PR DESCRIPTION
**Pull request by paulortman**
- [X] collector baseline
- [x] actor baseline
- [x] collector dev section support/selection
- [x] use the builtin "sha256" hash
- [x] track transitive and direct dependencies
- [x] add test for transitive vs direct dependencies
- [x] strip '==' in version strings in lockfiles
- [x] add custom setting to ignore [dev] section

### CODE REVIEW

**Comment by paulortman** [comment]
@davegaeddert I'm a little confused b/c these tests pass on my on development machine, but fail on Travis.  It's, at least in part, due to the fact that some machine/hardware/os stuff gets embedded in the Pipfile.lock file and things in Travis seems to be different.  I thought this docker thing was supposed to isolate the running tests in a fully repeatable environment...

**Comment by davegaeddert** [comment]
@paulortman are you seeing something like this as the difference?
```diff
8c8
<             "implementation_version": "3.6.1",
---
>             "implementation_version": "3.6.4",
15c15
<             "python_full_version": "3.6.1",
---
>             "python_full_version": "3.6.4",
```

I updated the Dockerfile to more specifically use 3.6.4... I do remember running into something like this before, where the cached layers on my end don't exactly match up with CI for some reason. I don't know much more than that though.

**Comment by davegaeddert** [comment]
Sorry, scratch that. That was the issue when I tried it on my machine but now I'm actually looking at Travis and seeing the other details.

**Comment by davegaeddert** [comment]
Ok, I changed it to use the builtin hash in Pipfile.lock. I'm almost tempted to try it that way for now rather than removing all _meta -- assuming they are using that hash for the same thing that we are of "this lockfile is outdated". Looks to me like the pipfile library makes that hash, and the host-environment-markers get added after the fact by pipenv and won't be included. Thoughts?

https://github.com/pypa/pipenv/blob/06ae7369ac3ae1165b28815bc91883565506426f/pipenv/vendor/pipfile/api.py#L85-L98

The docker platform_release and platform_version thing is odd/surprising, but it looks like we need to deal with that sort of problem anyway to get reproducible fingerprints across machines (at least within reason). I've avoided it so far, but we may be getting close to the point where we will need to have a python2.7 version of this component too. Not sure what other kinds of system requirement things like this we will run into, but at some point that's where the ability to "build your own container" thing will come back into play if necessary...?

**Comment by paulortman** [comment]
@davegaeddert I've cleaned this up a bit and things seem to be working and fairly straightforward, and don't quite know where to go from here.

1) Maybe some yml settings for how to handle dev sections of the Pipfile?
2) Perhaps more tests?
3) Update README with config examples

Your thoughts?

File: src/facade.py [hunk 1]
```diff
@@ -96,6 +97,23 @@ def dio_dependencies(self):
             }
         return dependencies
 
+    def fingerprint(self):
```
Comment (paulortman) [inline src/facade.py]: This is what I was thinking, but hadn't done yet. 👍

**Comment by paulortman** [comment]
closes #3

**Comment by paulortman** [comment]
@davegaeddert 

I'm feeling fairly good about this right now.  Time to depress me with your careful look.

File: src/models.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    def dio_dependencies(self, direct_dependencies=None):
+        "Return dependencies.io formatted list of lockfile dependencies"
+        dependencies = {}
+        for dep in self.dependencies():
+            dependencies[dep.key] = {
+                'source': dep.source,
+                'installed': {'name': self.strip_version_str(str(dep.specs))},
+            }
+            if direct_dependencies:
+                dependencies[dep.key]['is_transitive'] = True if dep.key not in direct_dependencies else False
```
Comment (paulortman) [inline src/models.py]: @davegaeddert 

I guess I'm too lazy to look this up, but just to be sure "is_transitive" == True if a dependency was not explicitly listed as a dependency; if it was pulled in by an explicit dependency as a requirement.  Right?

Comment (davegaeddert) [reply src/models.py]: Right. The terminology I'm operating on currently is "direct" vs "transitive" dependencies. Direct being the ones you asked for, transitive being everything else you get as a result.

File: Pipfile [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+url = "https://pypi.python.org/simple"
+verify_ssl = true
+name = "pypi"
+
+[requires]
+python_version = '3.6'
+
+[packages]
+
+dparse = {git = "https://github.com/paulortman/dparse", editable = true, ref = "b38113da6a5d1de97949ac358ba54ffee1399d3c"}
```
[the remark this answers is not in the document]
Comment (paulortman) [reply Pipfile]: I did and it was accepted overnight.  I'll update this before I merge.

File: src/collect.py [hunk 4]
```diff
@@ -1,99 +1,60 @@
...
 
-    # TODO decide how this is going to work across the system:
-    # - always a directory, filename in setttings if necessary
-    # - directory of file, doesn't matter
-    # - will /repo be prefixed (why does it need to be? should always be working in /repo so path should be relative... leading / was why)
-    requirements_path = sys.argv[1]
-    # requirements_filename = os.getenv('SETTING_REQUIREMENTS_FILENAME', 'requirements.txt')
-    # requirements_path = path.join(user_given_path_in_repo, requirements_filename)
+    # The first argument should be the manifest file
+    manifest_path = sys.argv[1]
```
[the remark this answers is not in the document]
Comment (paulortman) [reply src/collect.py]: I think the component actor/collector should be guaranteed to always get the same thing -- a path probably relative to /repo.

I think dependencies.io should take care of sanitizing the path (removing '..') and making it a relpath before passing it to the component.

[the remark this answers is not in the document]
Comment (davegaeddert) [reply src/collect.py]: I knew you would have a sensible answer to this.

**Pull request merged**